lock-api-objects/src/main/java/com/palantir/lock/watch/LockMetadata.java

```diff
@@ -0,0 +1,142 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.lock.watch.LockMetadata.Created;
+import com.palantir.lock.watch.LockMetadata.Deleted;
+import com.palantir.lock.watch.LockMetadata.Unchanged;
+import com.palantir.lock.watch.LockMetadata.Updated;
+import org.immutables.value.Value;
+import org.immutables.value.Value.Immutable;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = Unchanged.class, name = Unchanged.TYPE),
+    @JsonSubTypes.Type(value = Updated.class, name = Updated.TYPE),
+    @JsonSubTypes.Type(value = Deleted.class, name = Deleted.TYPE),
+    @JsonSubTypes.Type(value = Created.class, name = Created.TYPE),
+})
+public interface LockMetadata {
+    <T> T accept(Visitor<T> visitor);
+
+    static Unchanged unchanged() {
+        return ImmutableUnchanged.builder().build();
+    }
+
+    static Updated update(byte[] oldValue, byte[] newValue) {
+        return ImmutableUpdated.of(oldValue, newValue);
+    }
+
+    static Deleted delete(byte[] deletedValue) {
+        return ImmutableDeleted.of(deletedValue);
+    }
+
+    static Created create(byte[] createdValue) {
+        return ImmutableCreated.of(createdValue);
+    }
+
+    interface Visitor<T> {
+        T visit(Unchanged unchanged);
+
+        T visit(Updated updated);
+
+        T visit(Deleted deleted);
+
+        T visit(Created created);
+    }
+
+    /**
+     * LockMetadata event indicating that a value has definitely not been changed.
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableUnchanged.class)
+    @JsonDeserialize(as = ImmutableUnchanged.class)
+    @JsonTypeName(Unchanged.TYPE)
+    interface Unchanged extends LockMetadata {
+        String TYPE = "unchanged";
+
+        @Override
+        default <T> T accept(Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    /**
+     * LockMetadata event indicating that an existing value might have changed.
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableUpdated.class)
+    @JsonDeserialize(as = ImmutableUpdated.class)
+    @JsonTypeName(Updated.TYPE)
+    interface Updated extends LockMetadata {
+        String TYPE = "update";
+
+        @Value.Parameter
+        byte[] oldValue();
```

Comment:
Adding simply for discussion, I was also thinking maybe just shortening this to `old` and `new`, but I think keeping as-is makes more sense schematically. I had a similar class, https://github.com/palantir/atlasdb/blob/develop/atlasdb-workload-server-api/src/main/java/com/palantir/atlasdb/workload/invariant/MismatchedValue.java, but in this case the class type has `value` in it so it's a bit more obvious.

Reply:
Hm, omitting the suffix also saves us a few bytes during serialization so I am not opposed to the idea. Naming things `new` in Java can always lead to confusion. What do we think about using `before` and `after` instead? This would also clearly indicate that the different values refer to the points in time before and after the update operation took place.